File: 2734-lexicographically-smallest-string-after-substring-operation/2734-lexicographically-smallest-string-after-substring-operation.py
class Solution:
    def smallestString(self, s):
        s = list(s)  # Convert the string to a list of characters for efficient updates
        start = 0
        while start < len(s) and s[start] == 'a':
            start += 1  # Skip if 'a' is occurring as a substring from the starting
        if start == len(s):  # If reaches the end, we have to do exactly 1 operation, so just change the last character
            s[-1] = 'z'
        for i in range(start, len(s)):
            if s[i] == 'a':
                return ''.join(s)  # Convert the list back to a string
            else:
                s[i] = chr(ord(s[i]) - 1)  # Replace non-'a' characters with the previous character
        return ''.join(s)  # Convert the list back to a string

File: 2734-lexicographically-smallest-string-after-substring-operation/2734-lexicographically-smallest-string-after-substring-operation.cpp
class Solution {
public:
//     s = "leetcode"
// Output: "kddsbncd"
 
//     s = "acbbc"
// Output: "abaab"
    
//     s = "cbabc"
// Output: "baabc"
    
    string smallestString(string s) {
        
        int start=0;
        while(s[start]=='a')start++;    // skip if 'a' is occuring as substring from starting
        if(start==s.size()){                 // if reaches to last so we have to do exactly 1 operation so just change last one
            s[s.size()-1]='z';              
        }
        for(int i=start; i<s.size(); i++){
            if(s[i]=='a')return s;      
            else{
                s[i]=s[i]-1;   // else not 'a' means we can easily go prev to get smaller character
            }
        }
        return s;
    }
};